history/event.go

```diff
@@ -207,18 +209,43 @@ func (e *Event) UnmarshalJSON(in []byte) error {
 		if err := json.Unmarshal(wireEvent.MetadataBytes, &metadata); err != nil {
 			return err
 		}
-		e.Metadata = metadata
+		e.Metadata = &metadata
 		break
 	default:
 		if len(wireEvent.MetadataBytes) > 0 {
-			var metadata interface{}
+			var metadata UnknownEventMetadata
 			if err := json.Unmarshal(wireEvent.MetadataBytes, &metadata); err != nil {
 				return err
 			}
-			e.Metadata = metadata
+			e.Metadata = &metadata
 		}
 	}
 
 	// By default, leave the Event Metadata as map[string]interface{}
 	return nil
 }
+
+// EventMetadata is a type safety trick used to make sure that Metadata field
+// of Event is always a pointer, so that consumers can cast without being
+// concerned about encountering a value type instead. It works by virtue of the
+// fact that the method is only defined for pointer receivers; the actual
+// method chosen is entirely arbitary.
+type EventMetadata interface {
+	Type() string
+}
+
+func (cem *CommitEventMetadata) Type() string {
+	return EventCommit
+}
+
+func (cem *SyncEventMetadata) Type() string {
+	return EventSync
+}
+
+func (rem *ReleaseEventMetadata) Type() string {
+	return EventRelease
+}
+
+func (uem *UnknownEventMetadata) Type() string {
```

Comment:
See above re pointers to maps. I guess it looks odd if you don't uniformly use pointer receivers?